### PR TITLE
Dynamically disable CuArray scalar operations

### LINKED ISSUE #276: Dynamically disable CuArray scalar operations
https://github.com/climate-machine/Oceananigans.jl/blob/fc12f60371eb0df7ba3d088771ddc25c0f258de6/src/Oceananigans.jl#L154

@maleadt pointed out that this line might only execute when the package is precompiled so if we really want to enforce no scalar operations, we should do this in an `init.jl` file or in the `Model` constructor or something.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
Got bitten by CuArray scalar operations while debugging some Europa runs so this PR adds an `__init__()` function to the Oceananigans module that disables CuArray scalar operations once the module has been loaded and only allows scalar ops via the `CUDA.@allowscalar` macro.

Now when we accidentally perform scalar ops on CuArrays we should get an error, which is much better than having silent performance regressions which has happened multiple times in the past.

Two remaining problems:
1. `set!(u::Field, v::Number) = @. u.data = v` actually incurs scalar ops because `a::CuArray .= 1.5` does not but `a::OffsetArray{CuArray .= 1.5` does, so it's probably something we have to fix by adapting broadcasting over offsetarrays?
2. For similar broadcasting reasons I think
    ```julia
    data = cpudata(output)
    ds[name][:, :, :, time_index] = view(data, ow.slices[name]...)
    ```
    also incurs scalar ops during NetCDF output so I will try changing back to the old behavior where we used something like `	    
    ```julia
    data = interiorparent(output)
    !isa(output, Array) && (data = Array(data))
    ds[name][:, :, :, time_index] = view(data, ow.slices[name]...)
    ```

Resolves #276

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Ah so the dispatch on
```julia
@hascuda @inline cpudata(f::Field{X, Y, Z, <:CuArray}) where {X, Y, Z} =
    OffsetArray(Array(parent(f)), f.grid, location(f))
```
is wrong, it should be something like
```julia
Field{X, Y, Z, <:OffsetArray{T, D, CuArray} where {T, D}} where {X, Y, Z}
```
but still trying to get it to work.

Will add `cpudata` tests when I figure it out.

File: src/Fields/set!.jl [hunk 1]
```diff
@@ -52,7 +52,7 @@ function set!(Φ::NamedTuple; kwargs...)
     return nothing
 end
 
-set!(u::Field, v::Number) = @. u.data = v
+set!(u::Field, v::Number) = CUDA.@allowscalar @. u.data = v
```
Comment (glwagner) [inline src/Fields/set!.jl:55]: Why not use `u.data.parent .= v`?

File: test/test_halo_regions.jl [hunk 2]
```diff
@@ -26,7 +26,7 @@ function halo_regions_correctly_filled(arch, FT, Nx, Ny, Nz)
     grid = RegularCartesianGrid(FT, size=(Nx, Ny, Nz), extent=(Lx, Ly, Lz))
     field = CellField(FT, arch, grid)
 
-    interior(field) .= rand(FT, Nx, Ny, Nz)
+    CUDA.@allowscalar interior(field) .= rand(FT, Nx, Ny, Nz)
```
Comment (glwagner) [inline test/test_halo_regions.jl:29]: Should we use `set!` here instead?

File: src/Fields/set!.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/Fields/set!.jl:55]: Good idea, done.

File: test/test_halo_regions.jl (hunk 2 above)
Comment (ali-ramadhan) [reply test/test_halo_regions.jl]: Yup, done.

**Pull request merged**